Sources/PostgresNIO/Data/PostgresData+Range.swift

```diff
@@ -0,0 +1,244 @@
+import NIOCore
+
+// MARK: PostgresRangeBound
+
+public protocol PostgresRangeBound: PostgresDataConvertible, Comparable {
+    /// the Postgres range type associated with this bound
+    static var rangeType: PostgresDataType { get }
+    
+    /// If a Postgres range type has a well-defined step,
+    /// Postgres automatically converts it to a canonical form.
+    /// Types such as `int4range` get converted to upper-bound-exclusive.
+    /// This method is needed when converting an upper bound to inclusive.
+    var stepDown: (() -> Self)? { get }
```

Comment:
Why is that a property and not a simple method? Why should this be optional?

```swift
func decrement() -> Self
```

Reply:
I changed it to a method here: https://github.com/vapor/postgres-nio/pull/330/commits/1b3d009c04c6bab14af5ba48d402d1c4a9a86137. (I forget why I made it a property initially).

I also changed it from returning an optional value to making the function throw, which I think is clearer. The reason it can't always return a value is because some Postgres range types are "continuous" and don't have a well-defined step. This includes the built-in Postgres types numrange, tsrange, and tstzrange. Section 8.17.7 of the docs here explains it: https://www.postgresql.org/docs/current/rangetypes.html#RANGETYPES-DISCRETE.

I originally attempted to do something with Swift's `Strideable` but couldn't come up with a solution. Here's essentially what I would like to do in the `ClosedRange` conformance to `PostgresDecodable`, though this code isn't valid Swift:
```
if let strideable = upperBound as? Strideable {
    upperBound = strideable.advanced(by: -1)
}
```